Reinforce anonymity requirements and core use-cases (#33)
alxdavids/privacy-pass-ietf: draft-davidson-pp-protocol.md

      -
        ins: F. Valsorda
        org: Independent

>   TrustTokenAPI:

Reviewer: Probably pointing at the WICG repo as the canonical one makes sense, though maybe point it at the web.dev post until I get around to updating the WICG with the spec.

Author: I'll change it back, I think what is in the WICG repo works.

before being sent. We use the TLS presentation language {{RFC8446}} to
describe the structure of protocol data types and messages.

# Background

We discuss the core motivation behind the protocol along with the
guarantees and assumptions that we make in this document.

## Motivating use-cases

The Privacy Pass protocol was originally developed to provide anonymous

> authorization of users of anonymity-preserving tools on the Internet,
> such as Tor. In particular, the protocol allows clients to reveal

Reviewer 2: ```suggestion
authorization of Tor users. In particular, the protocol allows clients to reveal
```

authorization tokens that they have been issued without linking the
authorization to the actual issuance event. This means that the tokens
cannot be used to link the browsing patterns of clients that reveal
tokens.

> The Privacy Pass protocol has a number of use cases in practice. See

Reviewer 2: ```suggestion
Beyond this use case, the Privacy Pass protocol has a number of use cases in practice. See
```

{{DGSTV18}}, {{TrustTokenAPI}}, {{PrivateStorage}}, {{OpenPrivacy}}, and
{{Brave}} for examples.

## Anonymity and security guarantees

> As mentioned above, the protocol provides anonymity-preserving tokens
> for authorizing clients. Throughout this document, we use the terms

Reviewer 2: ```suggestion
Privacy Pass provides anonymity-preserving authorization tokens for clients. 
Throughout this document, we use the terms
```

"anonymous", "anonymous-preserving" and "anonymity" to refer to the core

> security guarantee of the protocol. This guarantee is the following
> statement.
> 
> - A token issued by a server key and subsequently redeemed is
>   indistinguishable from any other token issued under the same key.

Reviewer 2: ```suggestion
security guarantee of the protocol. Informally, this guarantee means that
any token issued by a server key and subsequently redeemed is indistinguishable 
from any other token issued under the same key.
```

Reviewer 3: Importantly, of course, this also requires that all clients are issued tokens under the same key for the same epoch when the key is valid (thereby actually ensuring the global anonymity set).

> We also require that clients cannot forge tokens, as otherwise Privacy
> Pass would have little value as an authorization protocol. This
> requirement is stated as follows.
> 
> - Any client that is issued `N` tokens under a given server key cannot
>   redeem more than `N` valid tokens.

Reviewer 2: ```suggestion
Privacy Pass also prohibits clients from forging tokens, as otherwise the protocol
would have little value as an authorization protocol. Informally, this means any client 
that is issued `N` tokens under a given server key cannot redeem more than `N` valid 
tokens.
```

before being sent. We use the TLS presentation language {{RFC8446}} to
describe the structure of protocol data types and messages.

# Background

We discuss the core motivation behind the protocol along with the
guarantees and assumptions that we make in this document.

## Motivating use-cases

The Privacy Pass protocol was originally developed to provide anonymous
authorization of users of anonymity-preserving tools on the Internet,
such as Tor and VPNs. In particular, the protocol allows clients to
reveal authorization tokens that they have been issued without linking
the authorization to the actual issuance event. This means that the
tokens cannot be used to link the browsing patterns of clients that
reveal tokens.

> The web performance company Cloudflare uses a form of Privacy Pass to
> issue tokens to users that have completed Internet challenges
> {{DGSTV18}}. This allows clients to redeem tokens instead of solving
> additional challenges in the future. Other use-cases have since arisen
> with similar core security goals: {{TrustTokenAPI}}, {{PrivateStorage}},
> {{OpenPrivacy}}, {{Brave}}.

Reviewer 2: ```suggestion
The Privacy Pass protocol has a number of use cases in practice. 
See {{DGSTV18}}, {{TrustTokenAPI}}, {{PrivateStorage}}, {{OpenPrivacy}}, 
and {{Brave}} for examples.
```

## Anonymity and security guarantees

As mentioned above, the protocol provides anonymity-preserving tokens
for authorizing clients. Throughout this document, we use the terms
"anonymous", "anonymous-preserving" and "anonymity" to refer to the core
security guarantee of the protocol. This guarantee is the following
statement.

> - Any client that is issued a token by a server key and subsequently
>   redeems it, remains anonymous within the set of all clients issued
>   tokens under the same key.

Reviewer 2: Focusing on the clients may be overly generic. What if clients were de-anonymized by IP address, for example? Perhaps:

```suggestion
- A token issued by a server key and subsequently redeemed is indistinguishable from any other 
  token issued under the same key.
```

> We also require a security guarantee for the server to maintain the
> utility of the authorization protocol.

Reviewer 2: ```suggestion
We also require that clients cannot forge tokens, as otherwise Privacy Pass would
have little value as an authorization protocol. This requirement is stated as follows.
```

- Any client that is issued `N` tokens under a given server key cannot
  redeem greater than `N` valid tokens.

> We discuss all of the protocol security requirements more thoroughly in
> {{sec-reqs}}.

Reviewer 2: ```suggestion
{{sec-reqs}} elaborates on these protocol anonymity and security requirements.
```

The client initialises a global storage system `store` that allows it
store the tokens that are received during issuance. The storage system
is a map of Issuer identifiers (`Issuer.id`) to arrays of stored tokens.

> We assume that the client knows the Issuer public key `pkI` ahead of

Reviewer 3: And of course that clients have a mechanism to verify their view of the Issuer public key is the same as all other clients.

Author: Added a link to the discussion about this in the architecture doc.
